### ISSUE TITLE
Allow to manually add NPM dependencies (peerDependencies, devDependencies and dependencies)

### ISSUE THREAD

**Issue by jjoslet** [vojtechhabarta/typescript-generator#633]
This is currently possible to add import declarations but there is no way to add NPM dependencies in the `package.json` file from the configuration.

In addition, the `package.json` model does not handle `peerDependencies`.

Here is an example useage:
1. Map `java.time.LocalDate` to JS Joda `LocalDate`
2. Import JS Joda package using `importDeclarations`
3. Add js joda package to `peerDependencies`, `devDependencies` and/or `dependencies.`

**Comment by vojtechhabarta** [comment]
I think we should add `npmDependencies` parameter to support use-case you mentioned.

But I would not add parameters for other types of dependencies unless there is a real need for it. In more complex situations it is always possible to enrich generated package.json file in next build step for example using `npm install` command with respective parameter.

**Comment by jjoslet** [comment]
Here's my company use cases if it can help you undesrtand why we want to use this feature.

We write µService with Spring Boot that provide REST API consumed by both Java and Angular frontend.

Since these pieces of software are mainly written by the same team, we manage to share the API model between the backend and the frontend.
So we write the API model in Java (annotated with Jackson) and generate it to TypeScript using your Maven Plugin.
The Maven model project build a Java package and a NPM package, both published to our company repository.

We have many application that follows that pattern.

Therefore, in our Angular common framework, we create tool to deserialize JSON and verify it against the model. We use the `reflect-metadata` package to decorate the generated model.
I write a plugin to your generator to generate the meta-data. These meta-data has a dependency to our meta-data API. This API is then a dependency of our framework and our models.

In that situation, this dependency has to be a `peerDependency` to avoid issues when the model and the framework communicate using our meta-data library.

Since a `peerDependency` is not automatically installed by the Node tools during development, this dependency must also be a `devDependency` to ease the development.

In the NPM world, a dependency must be a `peerDependency` if this dependency is part of the API. If it's an internal dependency, it must be a "normal" dependency.

Source: https://nodejs.org/en/blog/npm/peer-dependencies/

Please note that one cannot use npm install tool to add `peerDependency` in the package.json (https://docs.npmjs.com/cli/v6/commands/npm-install and https://github.com/npm/npm/pull/3994).

I see that there is also an `optionalDependencies` and a `bundleDependencies` in the package.json that can be useful for someone one day.

**Comment by vojtechhabarta** [comment]
Thanks for your use case and the link to a great explanation of peer dependencies. Just for curiosity: if you would put `reflect-metadata` into "normal" dependencies it should work most of the time because dependencies are installed flattened (top-level) in `node_modules` these days but it would break if two dependencies would declare different versions of `reflect-metadata` in their dependencies in which case it would not be installed top-level? Anyway I agree that peer dependency is the right solution.

**Comment by jjoslet** [comment]
You're right.

Suppose two packages `depA` and `depB` that both depend to `reflect-metadata` as a "normal" dependency and `my-project` uses these two dependencies.

If they both depend on the same version of `reflect-metadata`, there is no problem, it'll be installed at the root of `my-project`.

If they depend to different version of `reflect-metadata`, one `reflect-metadata` will be hosted at the root `node_modules` of `my-project`, the second will be hosted in the package specific `node_modules`. Something like that:
```
my-project
├── depA
├── reflect-metadata@1.0.0
└── depB
    └── reflect-metadata@2.0.0
```

This can then cause trouble with `instanceof`.

**Comment by vojtechhabarta** [comment]
Released in [2.30.840](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.30.840).
